torch/csrc/jit/passes/onnx/shape_type_inference.cpp

```diff
@@ -638,16 +652,211 @@ c10::optional<std::vector<int64_t>> GetValueFromListConstructNode(
       : c10::nullopt;
 }
 
+void ProcessBroadCastNode(Node* n) {
+  if (ConstantValueMap::HasShape(n->input(0)->debugName()) &&
+      ConstantValueMap::HasShape(n->input(1)->debugName())) {
+    auto input_shape_0 = ConstantValueMap::GetShape(n->input(0)->debugName());
+    auto input_shape_value_0 = input_shape_0.value().sizes();
+    auto input_shape_1 = ConstantValueMap::GetShape(n->input(1)->debugName());
+    auto input_shape_value_1 = input_shape_1.value().sizes();
+    size_t rank_0 = input_shape_value_0.value().size();
+    size_t rank_1 = input_shape_value_1.value().size();
+    size_t rank_max = std::max(rank_0, rank_1);
+    size_t rank_min = std::min(rank_0, rank_1);
+    std::vector<::c10::ShapeSymbol> final_shape;
+    final_shape.reserve(rank_max);
+    for (auto idx = 0; idx < rank_max; idx++) {
+      final_shape.emplace_back(::c10::ShapeSymbol::newSymbol());
+    }
+    for (auto idx = 0; idx < rank_min; idx++) {
+      auto is_static_0 =
+          input_shape_value_0.value()[rank_0 - 1 - idx].is_static();
+      auto is_static_1 =
+          input_shape_value_1.value()[rank_1 - 1 - idx].is_static();
+      if (is_static_0 && is_static_1) {
+        auto static_0_sz =
+            input_shape_value_0.value()[rank_0 - 1 - idx].static_size();
+        auto static_1_sz =
+            input_shape_value_1.value()[rank_1 - 1 - idx].static_size();
+        final_shape[rank_max - 1 - idx] = ::c10::ShapeSymbol::fromStaticSize(
+            std::max(static_0_sz, static_1_sz));
+      }
+    }
+
+    if (rank_0 < rank_1) {
+      for (auto idx = rank_min; idx < rank_max; idx++) {
+        auto shape_idx = rank_max - 1 - idx;
+        final_shape[shape_idx] = input_shape_value_1.value()[shape_idx];
+      }
+    } else {
+      for (auto idx = rank_min; idx < rank_max; idx++) {
+        auto shape_idx = rank_max - 1 - idx;
+        final_shape[shape_idx] = input_shape_value_0.value()[shape_idx];
+      }
+    }
+
+    UpdateShape(n->output(0), c10::SymbolicShape(final_shape));
+  }
+}
+
+void ProcessConcatNode(Node* n) {
+  int axis = n->i(attr::axis);
+  if (ConstantValueMap::HasRank(n->input(0)->debugName())) {
```

Comment:
Again, please assert that concat node n has at least 1 inputs.

Reply:
This is for a specific op, `Concat`. The assertion should be added in symbolic function if needed. Here we need assume that all the node are strictly following onnx spec.